File: src/python/posit.py
import math
import struct
from typing import override


class Posit:
    """
    Implementation of a Posit number.

    Implements parts of the [Posit standard](https://posithub.org/about)
    """

    def __init__(self, value: float|str = 0.0, n: int = 32):
        """
        Initialize a Posit number with the given value, precision, and exponent size.

        A posit is structured as follows:
        Sign Bit | Regime Bits | Exponent Bits | Fraction Bits
        n-1..n-2 |  n-3..n-k   |  n-k-1..n-m   |   n-m-1..0

        1. Sign Bit (1 bit): 
           0 for positive, 1 for negative. 
           Negative numbers are represented using Two's Complement.

        2. Regime Bits (Variable length):
           A run of identical bits terminated by an opposite bit. 
           Determines the regime value (k), representing a coarse scale factor of 16^k.

        3. Exponent Bits (Up to 2 bits):
           The next 2 bits represent the unsigned exponent (e) over the base 2.
           This field is truncated if the regime consumes too many bits.

        4. Fraction Bits (Remaining bits):
           The remaining bits form the fraction (f) of the significand 1.f.

        Args:
            value (float|str): The value of the Posit number,
                or the binary representation as a bit string.
            n (int): The precision of the Posit number, i.e. the number of total bits used.
        """
        if n < 2:
            raise ValueError("Posit precision n must be at least 2")

        self._bits: str = "0" * n
        """Bit sequence of the posit number"""

        self.BITWIDTH: int = n
        "Number of bits (n) used to represent the number."
        self.P_INT_MAX: int = int(math.ceil(2.0 ** math.floor(4 * (n - 3) / 5)))
        """The largest consecutive integer-valued posit value."""
        self.MIN_POS: float = 2.0 ** (-4 * n + 8)
        """The smallest positive posit value."""
        self.MAX_POS: float = 2.0 ** (4 * n - 8)
        """The largest positive posit value."""
        self._USEED: float = 16.0  # sepcified by posit standard v2

        if isinstance(value, (float, int)):
            if value != 0:
                self._float_to_posit(value)
        elif isinstance(value, str):
            self._set_bits(value.ljust(n, '0'))
        else:
            raise ValueError(value + " is not a float or string!")


    def _set_bits(self, bits: str):
        if not set(bits).issubset({'0', '1'}):
            raise ValueError(f"Error: '{bits}' is not a binary string!")
        if len(bits) != self.BITWIDTH:
            raise ValueError(f"Expected {self.BITWIDTH} bits, got {len(bits)}")
        self._bits = bits


    def is_zero(self) -> bool:
        """Tests if the value of this posit is equal to 0.0"""
        return self._bits == '0' * self.BITWIDTH


    def is_nar(self) -> bool:
        """Tests if the value of this posit is not-a-real (NaR)"""
        return self._bits == '1' + '0' * (self.BITWIDTH - 1)


    def clone(self) -> 'Posit':
        """Returns a deep copy of this instance."""
        res = Posit(n=self.BITWIDTH)
        res._bits = "".join(c for c in self._bits)
        return res


    def _float_to_posit(self, value: float):
        """Converts a 64-bit float into a n-bit posit"""
        n: int  = self.BITWIDTH

        # === Special Cases === #
        # Special case #1: zero
        if value == 0:
            self._set_bits("0" * n)
            return
        # Special case #2: nar
        if math.isnan(value) or math.isinf(value):
            self._set_bits("1" + "0" * (n - 1))
            return
        # Special case #3: float >= MAX_POS
        if value >= self.MAX_POS:
            self._set_bits('0' + '1' * (n - 1))
            return
        if abs(value) <= self.MIN_POS:
            self._set_bits('0' * (n - 1) + '1')
            return

        # === General Case === #
        # Pack to IEEE 754 double (64-bit)
        # [Sign (1)] [Exponent (11)] [Mantissa (52)]
        packed = struct.pack('>d', value)  # '>d' means Big-Endian Double
        float_bits = int.from_bytes(packed, 'big')
        ieee754_sign = (float_bits) >> 63 & 1  # Sign bit

        ieee754_exp = (float_bits >> 52) & 0x7FF  # (11 Bits)
        ieee754_mantissa = float_bits & 0xF_FFFF_FFFF_FFFF  # (52 Bits)

        # --- Extract Fraction (f) ---#
        if ieee754_exp == 0:  # - Subnormal: (-1)^s * 0.mant * 2^-1022
            # Normalize 0.mant to 1.f -> shift left until MSB is one
            bit_len = ieee754_mantissa.bit_length() - 1  # Leading 1 search
            total_exponent = bit_len - 1074  # 52 bits of fraction + 1022 exp

            f_int = ieee754_mantissa & ((1 << bit_len) - 1) # Shift mantissa left until MSB is 1
            f = format(f_int, f'0{bit_len}b') if bit_len > 0 else ""
        else:  # ---------------- Normal: (-1)^s * 1.mant * 2^(exp - 1023)
            total_exponent = ieee754_exp - 1023
            f = format(ieee754_mantissa, '052b')

        # --- Decompose total exponent into Regime (r) and Exponent (e) --- #
        # total_exponent = 4r + e
        r = total_exponent >> 2 # div 4
        e = total_exponent & 3  # mod 4

        # --- Encode parts into posit --- #
        self._encode_compnents(ieee754_sign, r, e, f)


    def _encode_compnents(self, s: int, r: int, e: int, f: str):
        """
        Encodes a Posit from its decoded components.

        Args:
            s (int): Sign (0 for positive, 1 for negative).
            r (int): Regime value (numerical exponent of useed).
            e (int): Exponent value (0-3).
            f (str): Fraction bits.
        """
        assert (0 <= s < 2) and (0 <= e < 4)
        n = self.BITWIDTH
        bit_str = '0'

        # === (n/n+x) bit Posit construction === #
        if r >= 0:
            k = r + 1
            bit_str += '1' * k + '0'
        else:
            k = -r
            bit_str += '0' * k + '1'
        bit_str += format(e, '02b')  # Binary string of length 2
        bit_str += f
        if len(bit_str) <= n:  # Value can be expressed as n-bit posit
            bit_str = bit_str.ljust(n, '0')
        else:  # === Posit rounding === #
            # Posit rounding works as follows:
            # let v be the n+1 bit posit currently stored in `posit_bits`
            # let u, w be the n bit posits, whose value surrounds v.
            u = bit_str[:n]  # next smaller n-bit posit
            # Because u, v and w have the same r, e incrementing u yields w.
            # As v has n+1 bits, LSB decides whether u or w are nearer.
            guard_bit = bit_str[n]  # n+1'th bit
            lsb = u[-1]
            above_midpoint = '1' in bit_str[n+1:]
            if guard_bit == '1' and (above_midpoint or lsb == '1'):
                # Above midpoint or round-to-even (Posit spec)
                bit_str = self._increment_bits(u)  #round up (v)
            else:
                bit_str = u  # Round down

        # === Apply sign bit === #
        if s == 0:
            self._set_bits(bit_str)
        else:
            self._set_bits(self._twos_complement(bit_str))


    def _get_components(self) -> tuple[int, int, int, str]:
        """
        Decodes the posit components:
            s (int): Sign (0 for positive, 1 for negative).
            r (int): Regime value (numerical exponent of useed).
            e (int): Exponent value (0-3).
            f (str): Fraction bits.

            Returns:
                tuple[int, int, int, str] of the form (s, r, e, f).
        """
        n = self.BITWIDTH
        bits = self._bits
        # === Read sign bit (s) === #
        s = 0 if self._bits[0] == '0' else 1
        if s == 1:
            bits = self._twos_complement(bits)
        #  === Parse regime (r) === #
        k = bits[1:].find('0' if bits[1] == '1' else '1')  # find run length
        if k == -1: # Regime is whole posit
            k = n-1
        r = -k if bits[1] == '0' else k - 1
        # === Read exponent (e) === #
        remaining = bits[2 + k:] # Stored in next 2 bits (spec)
        e = int(remaining[:2].ljust(2, '0'), 2)  # consider truncated exponent
        # === Read fraction (f) === #
        f = remaining[2:]
        return s, r, e, f


    def to_float(self) -> float:
        """Converts the value of this posit into a IEEE 754 float"""
        if self.is_zero():  # Special case 1
            return +0.0
        if self.is_nar():  # Special case 2
            return float('nan')
        s, r, e, f_str = self._get_components()
        f = 0.0  # Transform fixed point string into float
        if len(f_str) > 0:
            f = int(f_str.ljust(1, '0'), 2) / 2.0**len(f_str)
        float_val = (1 + f) * 2**(4*r + e)
        return float_val if s == 0 else -float_val


    def __eq__(self, other: 'Posit') -> bool:
        """
        Tests if two posits are numerically equal.
        Handles posits of different precisions (BITWIDTH).
        """
        if not isinstance(other, Posit):
            return False

        # === Special Cases === #
        if self.is_nar() and other.is_nar():
            return True # Posit spec
        if self.is_zero() and other.is_zero():
            return True
        if self.is_nar() or other.is_nar():
            return False

        # === Compare Components === #
        s1, r1, e1, f1 = self._get_components()
        s2, r2, e2, f2 = other._get_components()

        if s1 != s2 or r1 != r2 or e1 != e2:
            return False

        # === Compare Fractions === #
        max_len = max(len(f1), len(f2)) # "1" (0.5) should equal "100" (0.500)
        return f1.ljust(max_len, '0') == f2.ljust(max_len, '0')



    def __neg__(self) -> 'Posit':
        res = Posit(0, n=self.BITWIDTH)
        res._bits = self._twos_complement(self._bits)
        return res


    def __add__(self, other: 'Posit') -> 'Posit':
        """
        Adds this Posit to another Posit.

        Uses a unified-mantissa addition, meaning, components get massaged into:
        P = Mantissa * 2^Scale

        The Addition logic aligns the scales to the smaller common scale to preserve precision.
        Given P1 = M1 * 2^S1 and P2 = M2 * 2^S2 where S1 > S2:
        P1 becomes (M1 << (S1 - S2)) * 2^S2
        Then P1 + P2 = ((M1 << (S1 - S2)) + M2) * 2^S2
        """
        if not isinstance(other, Posit):
            raise ValueError(f"__add__: Expected Posit, got {type(other)}")
        if self.BITWIDTH != other.BITWIDTH:
            raise ValueError(f"Cannot add {self.BITWIDTH}-bit Posit with {other.BITWIDTH}-bit Posit!")

        # --- Handle Special Cases ---
        if self.is_nar() or other.is_nar():
            return Posit(n=self.BITWIDTH, value=float('nan'))
        if self.is_zero():
            return other.clone()
        if other.is_zero():
            return self.clone()

        # --- Calculate components --- #
        s_self, r_self, e_self, f_self = self._get_components()
        s_other, r_other, e_other, f_other = other._get_components()

        # --- Calculate Mantissa and Scale --- #
        mantissa_self, scale_self = self._unified_mantissa(r_self, e_self, f_self)
        mantissa_other, scale_other = self._unified_mantissa(r_other, e_other, f_other)

        # === Align to smaller, common scale === #
        if scale_self > scale_other:
            mantissa_self <<= (scale_self - scale_other)
            scale = scale_other
        else:
            mantissa_other <<= (scale_other - scale_self)
            scale = scale_self

        # === Perform addition === #
        if s_self == s_other:  # Same sign: Add magnitudes
            mantissa = mantissa_self + mantissa_other
            s = s_self
        else:  # Different signs: Subtract magnitudes
            if mantissa_self > mantissa_other:
                mantissa = mantissa_self - mantissa_other
                s = s_self
            elif mantissa_other > mantissa_self:
                mantissa = mantissa_other - mantissa_self
                s = s_other
            else:  # Exact cancellation
                return Posit(n=self.BITWIDTH)

        # --- Transform unified mantissa back into posit components --- #
        r, e, f = self._posit_parts(mantissa, scale)

        res = Posit(0, n=self.BITWIDTH)
        res._encode_compnents(s, r, e, f)
        return res


    def __mul__(self, other: 'Posit') -> 'Posit':
        """
        Multiplies this Posit by another Posit.

        Uses a unified-mantissa multiplication, meaning, components get massaed into:
        P = Mantissa * 2^Scale

        The Multiplication can then be rewritten as:
        (Mantissa1 * 2^Scale1) * (Mantissa2 * 2^Scale2) = (Mantissa1 * Mantissa2) * 2^(Scale1 + Scale2)
        """
        if not isinstance(other, Posit):
            raise ValueError(f"__mul__: Expected Posit, got {type(other)}")
        if self.BITWIDTH != other.BITWIDTH:
            raise ValueError(f"Cannot multiply {self.BITWIDTH}-bit Posit with {other.BITWIDTH}-bit Posit!")

        # === Handle Special Cases === #
        if self.is_nar() or other.is_nar():
            return Posit(float('nan'), n=self.BITWIDTH)
        if self.is_zero() or other.is_zero():
            return Posit(0.0, n=self.BITWIDTH)

        # --- Calculate components --- #
        s_self, r_self, e_self, f_self = self._get_components()
        s_other, r_other, e_other, f_other = other._get_components()

        # --- Calculate Mantissa and Scale --- #
        mantissa_self, scale_self = self._unified_mantissa(r_self, e_self, f_self)
        mantissa_other, scale_other = self._unified_mantissa(r_other, e_other, f_other)

        # === Perform multiplication === #
        mantissa = mantissa_self * mantissa_other
        scale = scale_self + scale_other
        sign = 1 if s_self != s_other else 0

        # --- Transform unified mantissa back into posit components --- #
        r, e, f = self._posit_parts(mantissa, scale)

        res = Posit(0, n=self.BITWIDTH)
        res._encode_compnents(sign, r, e, f)
        return res


    @override
    def __str__(self) -> str:
        return str(self.to_float()) + 'p'


    @override
    def __repr__(self) -> str:
        return f'Posit{self.BITWIDTH} ({self._bits})\n{self.to_float()}'


    @staticmethod
    def _twos_complement(bits: str) -> str:
        """Calculates the two's complement of a bit string of arbitrary length."""
        if "1" not in bits:
            return bits  # 0...0
        flipped = "".join(["1" if b == "0" else "0" for b in bits])
        return Posit._increment_bits(flipped)


    @staticmethod
    def _increment_bits(bits: str) -> str:
        """Adds 1 to a bit string of arbitrary length, handling overflow."""
        n = len(bits)
        val = int(bits, 2) + 1
        bin_val = format(val, f"0{n}b")
        return bin_val[-n:]  # Return last n bits in case of overflow


    @staticmethod
    def _unified_mantissa(r: int, e: int, f: str) -> tuple[int, int]:
        """
        Converts regime, exponent and fraction into mantissa and exponent:
        1.f * 2^(4r + e) -> m * 2^s

        While the fraction part has an implicit leading 1., the mantissa is
        entirely displayed and interpreted as an integer without a decimal
        point.

        Args:
            r (int): Regime value (numerical exponent of useed).
            e (int): Exponent value (0-3).
            f (str): Fraction bits.

        Returns:
            m, s (tuple[int, int]): Mantissa and scale (exponent). 
        """
        f_len = len(f)
        # Fraction currently is (1).XXX. Attach implicit one and interpret
        # as without decimal point (leftshift)
        m = (1 << f_len) | (int(f, 2) if f_len > 0 else 0)
        # Build unified mantissa (4r + e) and deduct leftshift of fraction
        s = (r * 4 + e) - f_len
        return m, s


    @staticmethod
    def _posit_parts(mantissa: int, scale: int) -> tuple[int, int, str]:
        """
        Converts a floating-point value composed of a mantissa and exponent
        into posit regime, exponent and fraction representation:
        m * 2^s -> 1.f * 2^(4r + e)

        While the mantissa is interpreted as an integer value, the fraction has
        an implicit 1 leading before the decimal point (1.f).

        Args:
            mantissa (int): An integer value
            scale (int): Integer exponent over the base 2.

        Returns:
            r, e, f (tuple[int, int, str]): Posit components (except sign)
        """
        mantissa_len = mantissa.bit_length() - 1  # Deduct implicit 1
        total_exponent = scale + mantissa_len  # Sum of scales + normalization shift
        r = total_exponent >> 2  # div 4
        e = total_exponent & 3   # mod 4
        f_int = mantissa & ((1 << mantissa_len) - 1)  # Discard leading 1
        f = format(f_int, f'0{mantissa_len}b') if mantissa_len > 0 else ""
        return r, e, f

File: src/cpp/posit.cc
#include "posit.h"

#include <algorithm>
#include <bit>
#include <cassert>
#include <cmath>
#include <compare>
#include <cstdint>

template <int N>
Posit<N>::Posit(double d) : bits(0) {
	static_assert(N <= 64);
	constexpr int n_ = N - 1;
	const double MAX_POS = std::pow(2.0, 4.0 * N - 8);
	const double MIN_POS = std::pow(2.0, -4.0 * N + 8);
	/* === Special Cases === */
	if (d == 0) return;  // Zero
	if (d == NAN || d == INFINITY || d == -INFINITY) {
		// bits |= 1ull << n_;  // Not a Real (NaR)
		*this = nar();
		return;
	}
	if (std::abs(d) >= MAX_POS) {
		bits |= (1ull << n_) - 1;  // Max. posit
		return;
	}
	if (std::abs(d) <= MIN_POS) {
		bits |= 1ull;  // Min. posit
		return;
	}
	/* === General Case === */
	auto float_bits = std::bit_cast<uint64_t>(d);
	int exponent = static_cast<int>((float_bits >> 52) & 0x7FFll);  // 11 Bits
	uint64_t mantissa = float_bits & 0xF'FFFF'FFFF'FFFFll;          // 52 Bits
	int total_exponent;
	uint64_t fraction;
	int frac_len{52};
	/* --- Extract fraction (f) */
	if (exponent == 0) {                                // Subnormal: (-1)^s * 0.mant * 2^-1022
		int msb_pos = 64 - std::countl_zero(mantissa);  // find pos of MSB in mantissa
		total_exponent = msb_pos - 1074;                // -1022 - 52 + shift
		fraction = mantissa & ((1ll << msb_pos) - 1);   // Convert subnormal to normal
		frac_len = msb_pos - 1;
	} else {  // Normal: (-1^s) * 1.man * 2^(exp - 1023)
		fraction = mantissa;
		total_exponent = exponent - 1023;
	}
	constexpr int STORAGE_BITS = sizeof(storage_t) * 8;
	int shift = frac_len - STORAGE_BITS;
	storage_t clipped_fraction;
	if (shift >= 0) {
		clipped_fraction = fraction >> shift;
	} else {
		clipped_fraction = fraction << abs(shift);
	}
	set_and_round(*this, (float_bits >> 63), total_exponent >> 2, total_exponent & 3, clipped_fraction);
}

template <int N>
constexpr std::strong_ordering Posit<N>::operator<=>(const Posit& other) const {
	if constexpr (SHIFT_ALIGN == 0) {  // Optimization for standard sizes (8, 16, 32, 64)
		return sbits <=> other.sbits;
	} else {  // For non-standard sizes, shift left to align sign bit, then compare.
		auto lhs = static_cast<sstorage_t>(bits << SHIFT_ALIGN);
		auto rhs = static_cast<sstorage_t>(other.bits << SHIFT_ALIGN);
		return lhs <=> rhs;
	}
}

template <int N>
Posit<N> Posit<N>::operator+=(const Posit<N>& other) {
	constexpr int BITS = std::numeric_limits<storage_t>::digits;
	constexpr storage_t MSB_MASK = static_cast<storage_t>(1) << (BITS - 1);

	//  Special Cases
	if (bits == 0) {
		*this = other;
		return *this;
	}
	if (other.bits == 0) return *this;
	if (*this == nar() || other == nar()) {
		*this = nar();
		return *this;
	}
	auto [r1, e1, f1] = this->get_components();
	auto [r2, e2, f2] = other.get_components();

	bool s1 = this->get_sign_bit();
	bool s2 = other.get_sign_bit();

	// Unify exponent: Exp = 4*r + e
	sstorage_t exp1 = static_cast<sstorage_t>(4) * static_cast<sstorage_t>(r1) + static_cast<sstorage_t>(e1);
	sstorage_t exp2 = static_cast<sstorage_t>(4) * static_cast<sstorage_t>(r2) + static_cast<sstorage_t>(e2);

	// Mantissas: Add Implicit 1 - Shift right logic 1 to place hidden bit at MSB.
	storage_t m1 = MSB_MASK | (f1 >> 1);
	storage_t m2 = MSB_MASK | (f2 >> 1);

	// Align exponents
	sstorage_t shift;
	if (exp1 < exp2) {
		std::swap(m1, m2);
		std::swap(s1, s2);
		std::swap(exp1, exp2);
	}
	shift = exp1 - exp2;
	storage_t sticky =  // Round up if all bits are 1
	    shift < BITS && (m2 & ((static_cast<storage_t>(1) << shift) - 1)) ? 1 : 0;
	m2 = (m2 >> shift) | sticky;

	// Add/Sub
	storage_t m_res = 0;
	int res_sign = s1;
	bool overflow = false;

	if (s1 == s2) {  // Handle signs
		m_res = m1 + m2;
		if (m_res < m1) overflow = true;
	} else {
		if (m1 >= m2) {
			m_res = m1 - m2;
		} else {
			m_res = m2 - m1;
			res_sign = !s1;
		}
	}

	if (m_res == 0) {
		bits = static_cast<storage_t>(0);
		return *this;
	}

	// Normalize (on overflow)
	if (overflow) {
		storage_t sticky = (m_res & 1);
		m_res = (m_res >> 1) | sticky;
		m_res |= MSB_MASK;  // The carry bit becomes MSB
		exp1++;
	} else {  // In case of subtraction check if mantissa needs to be aligned again
		int lz = std::countl_zero(m_res);
		if (lz > 0) {
			m_res <<= lz;
			exp1 -= lz;
		}
	}

	// Encode
	sstorage_t r_final = exp1 >> 2;
	sstorage_t e_final = exp1 & 3;
	storage_t f_final = m_res << 1;  // Extract Fraction (Remove hidden bit MSB)

	set_and_round(*this, res_sign, static_cast<int>(r_final), static_cast<int>(e_final), f_final);
	return *this;
}

template <int N>
Posit<N> Posit<N>::operator+(const Posit<N>& other) const {
	Posit res = *this;
	res += other;
	return res;
}

template <int N>
Posit<N> Posit<N>::operator*=(const Posit<N>& other) {
	constexpr int BITS = std::numeric_limits<storage_t>::digits;
	constexpr storage_t MSB_MASK = static_cast<storage_t>(1) << (BITS - 1);

	if (bits == 0 || other.bits == 0) {
		bits = static_cast<sstorage_t>(0);
		return *this;
	}
	if (*this == nar() || other == nar()) {
		*this = nar();
		return *this;
	}

	auto [r1, e1, f1] = this->get_components();
	auto [r2, e2, f2] = other.get_components();

	int s_res = (this->get_sign_bit() ^ other.get_sign_bit());

	// Formula:
	// P = (1.f1 * 2^E1) * (1.f2 * 2^E2)
	// E = (4r1 + e1) + (4r2 + e2)
	sstorage_t exp_res = (static_cast<sstorage_t>(4) * r1 + e1) + (4 * r2 + e2);

	// Attach implicit 1 to mantissas
	constexpr storage_t MSB_1 = static_cast<storage_t>(1) << (BITS - 1);
	storage_t m1 = MSB_1 | (f1 >> 1);
	storage_t m2 = MSB_1 | (f2 >> 1);

	// Multiply Mantissas: m1 * m2
	// Perform (sizeof(storage_t)*2)-bit multiplication 'in-place'
	constexpr int H = BITS / 2;
	constexpr storage_t LOW_MASK = (static_cast<storage_t>(1) << H) - 1;

	// (m1_up * H + m1_low) * (m2_up * H + m2_low) =
	//      (m1_low * m2_low)     % Ignored, precision will be lost anyway
	//    + (m1_low * m2_up * H)  % p_mid1
	//    + (m2_low * m1_up * H)  % p_mid2
	//    + (m1_up * m2_up)       % p_upper
	storage_t m1_lo = m1 & LOW_MASK;
	storage_t m1_up = m1 >> H;
	storage_t m2_lo = m2 & LOW_MASK;
	storage_t m2_up = m2 >> H;

	storage_t p_mid1 = m1_lo * m2_up;
	storage_t p_mid2 = m1_up * m2_lo;
	storage_t p_upper = m1_up * m2_up;

	storage_t p_mid = p_mid1 + p_mid2;  // Check if this statement overflows
	storage_t mid_carry = (p_mid < p_mid1) ? (static_cast<storage_t>(1) << H) : 0;

	storage_t m_prod = p_upper + (p_mid >> H) + mid_carry;

	// Normalize (Post-Normalization) - Result of 1.x * 1.y is in [1, 4).
	if ((m_prod & MSB_MASK) == 0) {
		m_prod <<= 1;  // shift 'left' to normalize (<< because right-alignment)
	} else {
		exp_res++;
	}

	storage_t f_final = m_prod & (~MSB_MASK);

	int r_final = exp_res >> 2;
	int e_final = exp_res & 3;

	f_final <<= 1;  // Remove implicit 1
	set_and_round(*this, s_res, static_cast<int>(r_final), static_cast<int>(e_final), f_final);
	return *this;
}

template <int N>
Posit<N> Posit<N>::operator*(const Posit<N>& other) const {
	Posit res = *this;
	res *= other;
	return res;
}

// template<int N>
// constexpr

/**
 * @brief Decodes the posit into its components: regime, exponent, and fraction.
 *
 * This function extracts the internal components of the posit value.
 * It handles negative values by taking the absolute value (two's complement)
 * before decoding.
 *
 * @tparam N The bit width of the posit.
 * @return A tuple containing:
 * - int r: The regime value.
 * - int e: The exponent value.
 * - storage_t f: The fraction bits. The fraction is returned right-aligned
 * in the integer container. Leading zeros are significant and implicit.
 */
template <int N>
std::tuple<int, int, typename Posit<N>::storage_t> Posit<N>::get_components() const {
	constexpr int BITS = std::numeric_limits<storage_t>::digits;
	auto abs = *this;
	if (get_sign_bit()) {
		abs = -abs;
	}
	int r, k, e{0};
	storage_t f{0};
	auto scratch = static_cast<storage_t>(abs.bits << SHIFT_ALIGN);  // Align posit bits with MSB
	scratch <<= 1;                                                   // Remove sign bit
	if (scratch & (static_cast<storage_t>(1) << (BITS - 1))) {       // Read direction
		k = std::countl_one(scratch);
		r = k - 1;
	} else {
		k = std::countl_zero(scratch);
		r = -k;
	}
	if (k + 1 >= BITS) return {r, e, f};
	scratch <<= (k + 1);  // Shift out regime
	e = static_cast<int>(scratch >> (BITS - 2));
	scratch <<= 2;  // Shift out exponent
	f = scratch;
	return {r, e, f};
}

template <int N>
double Posit<N>::to_double() {
	if (bits == 0) {
		return +0;
	}
	if (*this == nar()) {
		return NAN;
	}
	auto [r, e, f] = get_components();
	auto tmp = static_cast<double>(1ll << (N - 2));  // Make sure enough space left to cast to double
	double d = static_cast<double>(f >> 2) / tmp;    // No precision is lost, as sign+regime takes at least 3 bits
	d = (1.0 + d) * std::pow(2.0, 4.0 * r + e);
	return get_sign_bit() ? -d : d;
}

template <int N>
static void set_and_round(Posit<N>& posit, int s, int r, int e, typename Posit<N>::storage_t f) {
	// Automatically deduce the storage type U from the Posit definition
	using U = typename Posit<N>::storage_t;
	set_and_round_impl<N, U>(posit, s, r, e, f);
}

template <int N>
constexpr typename Posit<N>::storage_t increment(typename Posit<N>::storage_t val) {
	return ++val & Posit<N>::MASK;
}

template <int N>
constexpr typename Posit<N>::storage_t twos_complement(typename Posit<N>::storage_t val) {
	return ((~val) + 1) & Posit<N>::MASK;
}

/**
 * @brief Primary logic implementation.
 * Expects N (Posit Width) to match U (Storage Width) exactly.
 * Used for N = 8, 16, 32, 64.
 *
 * @param posit: Reference to the posit value
 * @param s: Sign bit, either false for 0 or true for 1
 * @param r: regime exponent
 * @param e: exponent, must be between 0 and 3
 * @param f: right-aligned fraction bits, so that the MSB is at position N-1<br>
 *           Example (32 bit): 0.625 | 31 31 30 29 28 ... 0 |<br>
 *                                   |  1  0  1  0  0 ... 0 |
 */
template <int N, typename U>
static void set_and_round_impl(Posit<N>& posit, bool s, int r, int e, U f) {
	// This implementation assumes the storage container fits the data exactly
	// to allow for native overflow behavior (simulating modulo arithmetic).

	assert(0 <= e && e < 4);
	constexpr auto SHIFT_ALIGN = Posit<N>::SHIFT_ALIGN;
	constexpr int BITS = std::numeric_limits<U>::digits;
	constexpr U ONE = static_cast<U>(1) << (BITS - 1);  // Leading one
	constexpr U ALL_ONES = ~static_cast<U>(0);
	r = std::clamp(r, -BITS, BITS - 1);

	int64_t k = (r >= 0) ? static_cast<int64_t>(r) + 1 : -static_cast<int64_t>(r);
	U regime_bits = r >= 0                        // Encode r in unary:
	                    ? ALL_ONES << (BITS - k)  // '1' * k + '0' (right aligned)
	                    : ONE >> k;               // '0' * k + '1' (right aligned)

	U posit_u{0};  // N bit posit
	// Posit u: | S | regime | exponent | fraction
	//    bits: | 1 | k + 1  |     2    |    m

	U posit_v{0};  // N+1 bit posit buffer (LSB is Guard, MSB is Regime Start)
	// Posit v: | regime | exponent | fraction
	//    bits: | k + 1  |     2    |    m
	posit_v |= regime_bits;
	posit_v |= static_cast<U>(e) << (BITS - k - 3);
	posit_v |= f >> (k + 3);              // k+1 (regime) + 2 (exponent)
	posit_u = posit_v >> (1 + BITS - N);  // Make space for sign

	// Check if rounding up is required
	constexpr U GUARD_SELECTOR = ONE >> (N - 1);
	bool lsb = (posit_u & static_cast<U>(1)) > 0;
	bool guard = (posit_v & GUARD_SELECTOR) > 0;
	if (guard) {                                                              // Might be above midpoint
		if (lsb                                                               // If lsb != 0: Round to next even posit
		    || (((static_cast<U>(1) << (SHIFT_ALIGN + k + 3)) - 1) & f) != 0  // If precision in the fraction was lost
		    || (e >> std::max((BITS - k - 3), 0l)) != 0) {                    // If precision in the exponent was lost
			posit_u = increment<N>(posit_u);
		}
	}
	// int pos = N - 1;      // Pointer to next unused space
	// bool above_midpoint{false};
	if (s) posit_u = twos_complement<N>(posit_u);  // Apply Sign
	posit.bits = posit_u;
}

// Explicitly instantiate for N=64 so the linker can find it
template struct Posit<64>;
template struct Posit<32>;
template struct Posit<16>;
template struct Posit<8>;
template struct Posit<4>;
